envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt

```diff
@@ -70,42 +68,88 @@ private fun getCommunicationMode(proto: Value?): CommunicationMode {
 }
 
 fun Value?.toStatusCodeFilter(accessLogFilterFactory: AccessLogFilterFactory):
-        AccessLogFilterSettings.StatusCodeFilterSettings? = this?.stringValue?.let {
+    AccessLogFilterSettings.StatusCodeFilterSettings? = this?.stringValue?.let {
     accessLogFilterFactory.parseStatusCodeFilter(it.toUpperCase())
 }
 
-private fun Value?.toOutgoing(properties: SnapshotProperties): Outgoing {
-    return Outgoing(
-        dependencies = this?.field("dependencies")?.list().orEmpty().map { it.toDependency(properties) }
+private class RawDependency(val name: String, val value: Value)
+
+fun Value?.toOutgoing(properties: SnapshotProperties): Outgoing {
+
+    val identifier = properties.outgoingPermissions.allServicesDependencies.identifier
+    val (serviceDependencies: List<RawDependency>,
+        domainDependencies: List<RawDependency>,
+        allServicesDependencies: Value?) = splitToDependencyTypes(
+        dependencies = this?.field("dependencies")?.list().orEmpty(),
+        identifier = identifier
     )
-}
 
-fun Value.toDependency(properties: SnapshotProperties = SnapshotProperties()): Dependency {
-    val service = this.field("service")?.stringValue
-    val domain = this.field("domain")?.stringValue
-    val handleInternalRedirect = this.field("handleInternalRedirect")?.boolValue
-        ?: properties.egress.handleInternalRedirect
-    val timeoutPolicy = this.field("timeoutPolicy")?.toOutgoingTimeoutPolicy(properties)
-        ?: Outgoing.TimeoutPolicy(
+    val defaultSettingsFromProperties = DependencySettings(
+        handleInternalRedirect = properties.egress.handleInternalRedirect,
+        timeoutPolicy = Outgoing.TimeoutPolicy(
             idleTimeout = Durations.fromMillis(properties.egress.commonHttp.idleTimeout.toMillis()),
             requestTimeout = Durations.fromMillis(properties.egress.commonHttp.requestTimeout.toMillis())
         )
-    val rewriteHostHeader = this.field("rewriteHostHeader")?.boolValue ?: false
-
-    val settings = DependencySettings(handleInternalRedirect, timeoutPolicy, rewriteHostHeader)
+    )
+    val allServicesDefaultSettings = allServicesDependencies.toSettings(defaultSettingsFromProperties)
+    return Outgoing(
+        serviceDependencies = serviceDependencies.map {
+            ServiceDependency(it.name, it.value.toSettings(allServicesDefaultSettings))
+        },
+        domainDependencies = domainDependencies.map {
+            DomainDependency(it.name, it.value.toSettings(defaultSettingsFromProperties))
+        },
+        defaultServiceSettings = allServicesDefaultSettings,
+        allServicesDependencies = allServicesDependencies != null
+    )
+}
 
-    return when {
-        service == null && domain == null || service != null && domain != null ->
-            throw NodeMetadataValidationException(
-                "Define either 'service' or 'domain' as an outgoing dependency"
+@Suppress("ThrowsCount")
+private fun splitToDependencyTypes(
```

Comment:
I think this function tries to do too much. Would it make more sense to split it into validation & extraction?

Reply:
i'm trying to refactor it

Reply:
I refactored this function. Please let me know what you think